Implement the `i18n.getMessage()` API which returns the localized string for a specified message.

kiaraarose opened the pull request:
#### 272d269ad268b83873c56b0a591e33c4471cebbb
<pre>
Implement the `i18n.getMessage()` API which returns the localized string for a specified message.
<a href="https://webkit.org/b/246488">https://webkit.org/b/246488</a>
rdar://102720462 (Web Extensions in WebKit: Add support for i18n(Localization) in WebKit)

Reviewed by Timothy Hatcher.

This patch adds support for i18n.getMessage(). This was done by storing the _WKWebExtensionLocalization
object for the extension onto the WebExtensionContextProxy, so that it could be used in the WebProcess
to return the localized string for the specified message name.

* Source/WebKit/Modules/OSX_Private.modulemap:
* Source/WebKit/Modules/iOS_Private.modulemap:
* Source/WebKit/Shared/Extensions/WebExtensionContextParameters.h:
* Source/WebKit/Shared/Extensions/WebExtensionContextParameters.serialization.in:
* Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.h:
* Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.mm:
(-[_WKWebExtensionLocalization initWithWebExtension:]):
We should use the resource methods on the WebExtension object to create the localized dictionary
instead of the using the bundleURL to find the resources on disk.

(-[_WKWebExtensionLocalization
initWithRegionalLocalization:languageLocalization:defaultLocalization:withBestLocale:uniqueIdentifier:]):
Removed nil check that  caused the uniqueIdentifier to never be set.
Fixed a bug where the regional dictionary was never being used/merged to the localizedDictionary.

(-[_WKWebExtensionLocalization localizedStringForKey:withPlaceholders:]):
(-[_WKWebExtensionLocalization _localizationDictionaryForWebExtension:withLocale:]):
(-[_WKWebExtensionLocalization _predefinedMessagesForLocale:]):
(localizationDictionaryAtURL): Deleted.
(-[_WKWebExtensionLocalization initWithBundleURL:defaultLocale:uniqueIdentifier:]): Deleted.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::manifest):
(WebKit::WebExtension::serializeLocalization):
(WebKit::WebExtension::resourceDataForPath):
Add support for NSDictionary.

* Source/WebKit/UIProcess/Extensions/WebExtension.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.cpp:
(WebKit::WebExtensionContext::parameters const):
Add localization dictionary to the parameters struct.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
Remove _WKWebExtensionLocalization from the Private header.

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPILocalizationCocoa.mm:
(WebKit::WebExtensionAPILocalization::getMessage):
(WebKit::WebExtensionAPILocalization::getUILanguage):
(WebKit::WebExtensionAPILocalization::getAcceptLanguages):

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm:
(WebKit::WebExtensionAPIPermissions::getAll):
(WebKit::WebExtensionAPIPermissions::contains):
(WebKit::WebExtensionAPIPermissions::request):
(WebKit::WebExtensionAPIPermissions::remove):

* Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionContextProxyCocoa.mm:
(WebKit::WebExtensionContextProxy::parseJSON):
(WebKit::WebExtensionContextProxy::parseLocalization):
(WebKit::WebExtensionContextProxy::parseManifest): Deleted.

* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.cpp:
(WebKit::WebExtensionContextProxy::getOrCreate):

* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPILocalization.mm: Added.
(TestWebKitAPI::localeStringInWebExtensionFormat):
(TestWebKitAPI::TEST):

* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.h:
(TestWebKitAPI::Util::constructJSArrayOfStrings):
A &quot;stringify&quot; helper method that turns a NSArray of elements into a string.

Canonical link: <a href="https://commits.webkit.org/267674@main">https://commits.webkit.org/267674@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9186cbbb6343acb8d58512402c9af8cc8fc4b94f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17371 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17695 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/18205 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/19157 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/16254 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17567 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20971 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17837 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18422 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17575 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17909 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/15106 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19975 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/15149 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15803 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/22465 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/16148 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15972 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20292 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16551 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/14045 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15697 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/20067 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2119 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16387 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kiaraarose commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionWindow.mm:
```diff
@@ -31,9 +31,11 @@
 #import "TestWebExtensionsDelegate.h"
 #import <WebKit/WKFoundation.h>
 #import <WebKit/_WKWebExtensionContextPrivate.h>
+#import <WebKit/_WKWebExtensionControllerPrivate.h>
```
was getting build errors in TestWebKitAPI so I added these imports

xeenon replied:
Yeah that happens when adding new files to the Sources.txt file. It means the file was missing the include but got it for free by being combined with another file before, and that file is no longer grouped with the same files due to the added file.

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm:
```diff
@@ -48,8 +48,6 @@
     // Documentation: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/permissions/getAll
 
-    RELEASE_LOG(Extensions, "permissions.getAll()");
-
```
i saw these got removed in one of your [PRs](https://github.com/WebKit/WebKit/pull/16854/commits/71338e8db0ce777a17016df4b4367a61f9261449#diff-c463bc70aecf160d7a810c45526f6f2f6b04bb4b80e0caab2c801056270e5f09) @xeenon ... not sure how they re-appeared

xeenon replied:
Odd. Yeah the code generator does this now so removing them is correct.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPILocalization.mm:
```diff
@@ -0,0 +1,324 @@
+        @"text end": @"Text direction is left",
+    };
+
+    localization = [[_WKWebExtensionLocalization alloc] initWithRegionalLocalization:@{ } languageLocalization:@{ } defaultLocalization:@{ } withBestLocale:@"ar_SA" uniqueIdentifier:uuidString];
```
This is a memory leak, you should either use `RetainPtr` or enable ARC for this file similar to `WebExtensionUtilities.mm`.

kiaraarose replied:
Good catch, will add ARC for this file.

pxlcoder commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -220,6 +220,11 @@
 
+Ref<API::Data> WebExtension::serializeLocalization()
+{
+    return API::Data::createWithoutCopying([NSJSONSerialization dataWithJSONObject:m_localization.get().localizationDictionary options:0 error:nullptr]);
```
```suggestion
    return API::Data::createWithoutCopying([NSJSONSerialization dataWithJSONObject:[m_localization localizationDictionary] options:0 error:nullptr]);
```

xeenon replied:
We prefer dot-syntax for properties.

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionContextParameters.serialization.in:
```diff
@@ -27,6 +27,7 @@ struct WebKit::WebExtensionContextParameters {
 
     URL baseURL;
     String uniqueIdentifier;
+    Ref<API::Data> localization;
```
Maybe `localizationJSON` to match `manifestJSON`? Or rename the `manifestJSON` to `manifest`?

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPILocalizationCocoa.mm:
```diff
@@ -51,8 +52,9 @@
-    // FIXME: <https://webkit.org/b/246488> Retrieve localized dictionary from the WebExtensionContextProxy.
-    _WKWebExtensionLocalization *localizedDictionary;
+    NSDictionary *localizationDictionary = extensionContext().localizationDictionary();
+    _WKWebExtensionLocalization *localization = [[_WKWebExtensionLocalization alloc] initWithRegionalLocalization:localizationDictionary languageLocalization:nil defaultLocalization:nil withBestLocale:nil uniqueIdentifier:extensionContext().uniqueIdentifier()];
```
It would be better to store `_WKWebExtensionLocalization` on the context, and access it here instead of the dictionary so we are not making `_WKWebExtensionLocalization` every time.

xeenon reviewed (changes requested):
Change `_WKWebExtensionLocalization` to be a member of the WebExtensionContextProxy instead of the dictionary.

kiaraarose commented on Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.mm:
```diff
@@ -32,16 +32,18 @@
-NSString * const _WKLocalizationDictionaryDescriptionKey = @"description";
-NSString * const _WKLocalizationDictionaryPlaceholdersKey = @"placeholders";
-
 static NSString * const localizationDictionaryJSONPathFormat = @"_locales/%@/messages.json";
```
`pathToJSONFile `

kiaraarose commented on Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
```diff
@@ -67,11 +68,15 @@ class WebExtensionContextProxy final : public RefCounted<WebExtensionContextProx
 
+    _WKWebExtensionLocalization *localization() { return m_localization.get(); }
+
+    static RetainPtr<_WKWebExtensionLocalization> parseLocalization(API::Data&);
```
`parseLocalizationJSON`

kiaraarose commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionWindow.mm:
```diff
@@ -38,6 +38,12 @@
 #import <WebKit/_WKWebExtensionPrivate.h>
 #import <WebKit/_WKWebExtensionWindow.h>
 
+@interface TestWindow : NSObject <_WKWebExtensionWindow>
```
i think this is a merge conflict error. will remove

xeenon replied:
Yeah I recently removed this class.

kiaraarose commented on Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.mm:
```diff
@@ -312,9 +300,9 @@ - (NSString *)_stringByReplacingPositionalPlaceholdersInString:(NSString *)strin
 #else
 
-- (instancetype)initWithBundleURL:(NSURL *)bundleURL defaultLocale:(NSString *)defaultLocaleString uniqueIdentifier:(NSString *)uniqueIdentifier
+- (instancetype)initWithWebExtension:(_WKWebExtension *)extension
```
`WebExtension&` to resolve build errors

xeenon commented on Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.mm:
```diff
@@ -241,29 +217,41 @@ - (NSString *)localizedStringForString:(NSString *)sourceString
+    NSString *path = [NSString stringWithFormat:localizationDictionaryJSONPathFormat, locale];
+    NSData *data = [NSData dataWithData:webExtension.resourceDataForPath(path)];
+    if (!data) {
+        RELEASE_LOG_DEBUG(Extensions, "Could not find localization for %{public}@ for extension %{private}@", locale, webExtension.displayName());
```
`webExtension.resourceDataForPath()` already logs this, can just return nil.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPILocalization.mm:
```diff
@@ -0,0 +1,268 @@
+        @"browser.test.assertEq(await browser.i18n.getMessage('@@bidi_reversed_dir'), reversedTextDirection)",
+        @"browser.test.assertEq(await browser.i18n.getMessage('@@bidi_start_edge'), startEdge)",
+        @"browser.test.assertEq(await browser.i18n.getMessage('@@bidi_end_edge'), endEdge)",
+        @"browser.test.assertEq(await browser.i18n.getMessage('unknown_message'), '')",
```
Shouldn't need `await` for `getMessage()` since it is a sync function that returns a result.

kiaraarose committed:
Implement the `i18n.getMessage()` API which returns the localized string for a specified message.
https://webkit.org/b/246488
rdar://102720462 (Web Extensions in WebKit: Add support for i18n(Localization) in WebKit)

Reviewed by Timothy Hatcher.

This patch adds support for i18n.getMessage(). This was done by storing the _WKWebExtensionLocalization
object for the extension onto the WebExtensionContextProxy, so that it could be used in the WebProcess
to return the localized string for the specified message name.

* Source/WebKit/Modules/OSX_Private.modulemap:
* Source/WebKit/Modules/iOS_Private.modulemap:
* Source/WebKit/Shared/Extensions/WebExtensionContextParameters.h:
* Source/WebKit/Shared/Extensions/WebExtensionContextParameters.serialization.in:
* Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.h:
* Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.mm:
(-[_WKWebExtensionLocalization initWithWebExtension:]):
We should use the resource methods on the WebExtension object to create the localized dictionary
instead of the using the bundleURL to find the resources on disk.

(-[_WKWebExtensionLocalization
initWithRegionalLocalization:languageLocalization:defaultLocalization:withBestLocale:uniqueIdentifier:]):
Removed nil check that  caused the uniqueIdentifier to never be set.
Fixed a bug where the regional dictionary was never being used/merged to the localizedDictionary.

(-[_WKWebExtensionLocalization localizedStringForKey:withPlaceholders:]):
(-[_WKWebExtensionLocalization _localizationDictionaryForWebExtension:withLocale:]):
(-[_WKWebExtensionLocalization _predefinedMessagesForLocale:]):
(localizationDictionaryAtURL): Deleted.
(-[_WKWebExtensionLocalization initWithBundleURL:defaultLocale:uniqueIdentifier:]): Deleted.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::manifest):
(WebKit::WebExtension::serializeLocalization):
(WebKit::WebExtension::resourceDataForPath):
Add support for NSDictionary.

* Source/WebKit/UIProcess/Extensions/WebExtension.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.cpp:
(WebKit::WebExtensionContext::parameters const):
Add localization dictionary to the parameters struct.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
Remove _WKWebExtensionLocalization from the Private header.

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPILocalizationCocoa.mm:
(WebKit::WebExtensionAPILocalization::getMessage):
(WebKit::WebExtensionAPILocalization::getUILanguage):
(WebKit::WebExtensionAPILocalization::getAcceptLanguages):

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm:
(WebKit::WebExtensionAPIPermissions::getAll):
(WebKit::WebExtensionAPIPermissions::contains):
(WebKit::WebExtensionAPIPermissions::request):
(WebKit::WebExtensionAPIPermissions::remove):

* Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionContextProxyCocoa.mm:
(WebKit::WebExtensionContextProxy::parseJSON):
(WebKit::WebExtensionContextProxy::parseLocalization):
(WebKit::WebExtensionContextProxy::parseManifest): Deleted.

* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.cpp:
(WebKit::WebExtensionContextProxy::getOrCreate):

* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPILocalization.mm: Added.
(TestWebKitAPI::localeStringInWebExtensionFormat):
(TestWebKitAPI::TEST):

* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.h:
(TestWebKitAPI::Util::constructJSArrayOfStrings):
A "stringify" helper method that turns a NSArray of elements into a string.

Canonical link: https://commits.webkit.org/267674@main
